finality-aleph/src/new_network/connection_manager/service.rs

```diff
@@ -0,0 +1,662 @@
+        if let Some(handler) = self
+            .sessions
+            .get(&session_id)
+            .map(|session| &session.handler)
+        {
+            let to_send = NetworkData::Data(message, session_id);
+            match recipient {
+                Recipient::Everyone => (0..handler.node_count().0)
+                    .map(NodeIndex)
+                    .flat_map(|node_id| handler.peer_id(&node_id))
+                    .map(|peer_id| {
+                        (
+                            to_send.clone(),
+                            DataCommand::SendTo(peer_id, Protocol::Validator),
+                        )
+                    })
+                    .collect(),
+                Recipient::Node(node_id) => handler
+                    .peer_id(&node_id)
+                    .into_iter()
+                    .map(|peer_id| {
+                        (
+                            to_send.clone(),
+                            DataCommand::SendTo(peer_id, Protocol::Validator),
+                        )
+                    })
+                    .collect(),
+            }
+        } else {
+            Vec::new()
+        }
+    }
+
+    /// Handle a discovery message.
+    /// Returns a command possibly changing what we should stay connected to and a list of data to
+    /// be sent over the network.
+    pub fn on_discovery_message(
+        &mut self,
+        message: DiscoveryMessage,
+    ) -> (
+        Option<ConnectionCommand>,
+        Vec<(NetworkData<D>, DataCommand)>,
+    ) {
+        let session_id = message.session_id();
+        match self.sessions.get_mut(&session_id) {
+            Some(Session {
+                handler, discovery, ..
+            }) => {
+                let (addresses, responses) = discovery.handle_message(message, handler);
+                let maybe_command = match addresses.is_empty() {
+                    false => {
+                        self.connections
+                            .add_peers(session_id, addresses.iter().flat_map(get_peer_id));
+                        Some(ConnectionCommand::AddReserved(
+                            addresses.into_iter().map(|address| address.0).collect(),
+                        ))
+                    }
+                    true => None,
+                };
+                (
+                    maybe_command,
+                    responses.into_iter().map(Self::network_message).collect(),
+                )
+            }
+            None => {
+                debug!(target: "aleph-network", "Received message from unknown session: {:?}", message);
+                (None, Vec::new())
+            }
+        }
+    }
+
+    /// Returns a data sink for data received that is associated with the given session.
+    pub fn get_session_channel(&self, session_id: &SessionId) -> Option<&mpsc::UnboundedSender<D>> {
+        self.sessions
+            .get(session_id)
+            .map(|session| session.data_for_user.as_ref())
```

Comment:
is there a reason for a ref and no clone here?

Reply:
You only call this when pushing a single value, so cloning seems wasteful.